webapp/src/Service/AssetUpdateService.php

```diff
@@ -0,0 +1,79 @@
+<?php declare(strict_types=1);
+
+namespace App\Service;
+
+use App\Entity\BaseApiEntity;
+use App\Entity\Contest;
+use App\Entity\Team;
+use App\Entity\TeamAffiliation;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\HttpFoundation\File\UploadedFile;
+
+class AssetUpdateService
+{
+    /**
+     * @var DOMJudgeService
+     */
+    protected $dj;
+
+    /**
+     * @var EventLogService
+     */
+    protected $eventLog;
+
+    public function __construct(DOMJudgeService $dj, EventLogService $eventLog)
+    {
+        $this->dj = $dj;
+        $this->eventLog = $eventLog;
+    }
+
+    /**
+     * The entities and asset fields supported by this subscriber
+     *
+     * The entities are assumed to have a property called <asset>File and clear<Asset>,
+     * where <asset> is the name of the asset in this array and <Asset> is the same but
+     * with the first letter uppercased. <asset>File should be an `UploadedFile` and
+     * clear<Asset> should be a boolean.
+     */
+    protected const SUPPORTED_ASSETS = [
+        Team::class            => ['photo'],
+        TeamAffiliation::class => ['logo'],
+        Contest::class         => ['banner'],
+    ];
+
+    protected function getAssetFields($entity): array
+    {
+        $class = get_class($entity);
+        if (strpos($class, 'Proxies\\__CG__\\') === 0) {
+            $class = substr($class, strlen('Proxies\\__CG__\\'));
+        }
+        return static::SUPPORTED_ASSETS[$class] ?? [];
+    }
+
+    /**
+     * Update assets for the given entity
+     */
+    public function updateAssets(BaseApiEntity &$entity): void
+    {
+        $assetFields = $this->getAssetFields($entity);
+        if (!$assetFields) {
+            return;
+        }
+
+        $fs = new Filesystem();
+
+        foreach ($assetFields as $assetField) {
+            $assetPath = $this->dj->fullAssetPath($entity, $assetField, $this->eventLog->externalIdFieldForEntity($entity) !== null);
+            $clearMethod = sprintf('isClear%s', ucfirst($assetField));
+            $shouldClear = call_user_func([$entity, $clearMethod]);
```

Comment:
Can we solve this in a different way? Doing such dynamic function calls makes the code less discoverable and I'd prefer to avoid this.

Reply:
I welcome ideas. I could put a `switch` statement here that loops over all types, basically getting rid of the constant. But then you have logic all over the place, also doesn't sound ideal.

Another option would be to add a method to each of the three classes (Team, TeamAffiliation and Contest) with

```
public function isClearAsset(string $field) {}
// and
public function getAssetFile(string $field) {}
```
which both have a switch statement. Maybe that is better?

Reply:
I don't have a perfect overview, but I think adding this latter `public function isClearAsset(string $field)` and `getAssetFile` might be nicer. I'm wondering what others think @thijskh @meisterT ?

Reply:
I also think it looks better. I will give it a try tomorrow unless someone objects

Reply:
The `getApiId` fix is nice BTW: that cleans things up a lot!

Reply:
In the end it still has a magic method call but then only in one place. I like it as well